check/src/typecheck.rs

```diff
@@ -1312,13 +1313,15 @@ impl<'a> Typecheck<'a> {
 fn with_pattern_types<F>(fields: &[(Symbol, Option<Symbol>)], typ: &TcType, mut f: F)
     where F: FnMut(&Symbol, &Option<Symbol>, &TcType)
 {
-    if let Type::Record { fields: ref field_types, .. } = **typ {
-        for field in fields {
-            // If the field in the pattern does not exist (undefined field error) then skip it as
-            // the error itself will already have been reported
-            if let Some(associated_type) = field_types.iter()
-                .find(|type_field| type_field.name.name_eq(&field.0)) {
-                f(&field.0, &field.1, &associated_type.typ);
+    if let Type::Record { ref row, .. } = **typ {
+        if let Type::ExtendRow { fields: ref field_types, .. } = **row {
+            for field in fields {
+                // If the field in the pattern does not exist (undefined field error) then skip it as
+                // the error itself will already have been reported
+                if let Some(associated_type) = field_types.iter()
+                    .find(|type_field| type_field.name.name_eq(&field.0)) {
+                    f(&field.0, &field.1, &associated_type.typ);
+                }
```

Comment:
I don't think we need to check that exactly but there is definitely a bug here since there could be a nested `ExtendRow { rest: ExtendRow { .. } }` and I only check the outermost row. Need to change to to use `Type::field_iter` as I use in most places.